field/src/types.rs

```diff
@@ -556,6 +556,16 @@ pub trait PrimeField64: PrimeField + Field64 {
 
     fn to_noncanonical_u64(&self) -> u64;
 
+    /// The conversion to a canonical i64 is mostly useful for debugging,
+    /// as it tries to recover 'negative' values for better readability.
+    fn to_canonical_i64(&self) -> i64 {
+        i64::try_from(self.to_canonical_u64()).unwrap_or_else(|_| {
```

Comment:
I'm a bit confused about this method. Do we want to interpret integers in the range `[0, M/2[` where `M` is the modulus as being canonical positive integers, and `[M/2, M[` being negative integers?
If so, this is slightly incorrect, as for instance with Goldilocks, the negative range spans from `2^63 - 2^16` to `2^64 - 2^32`, but the first `2^16` elements of the range would appear as positive elements.

Reply:
I mostly just want 'small' negative numbers like -1, or -100 to survive a round-trip into field elements and back.

And I want the result to be independent of whether the field element has had `to_canonical` applied to it or not.

So the implied semantics are basically: put the canonical representation of the number into a non-negative `i64`,  unless that doesn't fit; in that case, give me the negative `i64`.

However you are right that splitting the range somewhere around `M/2` instead of `i64::MAX` might be less confusing and easier to explain and would also give a reasonable answers (that survives round-tripping in the sense mentioned above) for fields with `M < i64::MAX`.

Reply:
Sorry I had missed it. Yes I'd tend to prefer the `M/2` route for clarity.